### DISCUSSION TITLE
1.8.0

### DISCUSSION THREAD

**Discussion by MariusWirtz** [cubewise-code/tm1py#619] [General]
# Highlights

- Support MDX properties in `execute_mdx_dataframe`.
This allows to query data and attribute data with a single query into one shared data frame.

``` python
from TM1py import TM1Service

with TM1Service(address="", port=12354, user="admin", password="apple", ssl=True) as tm1:
    mdx = """
    SELECT
    {Tm1SubsetAll([Region])} PROPERTIES [Region].[Currency] ON COLUMNS,
    {Tm1SubsetAll([Year])} ON ROWS
    FROM [Sales]
    """

    data = tm1.cells.execute_mdx_dataframe(mdx, include_attributes=True)
```

|    | Year | Region | Currency  |   Value |
|---:|:----------|:------|:----------------------|--------:|
|  0 | 2021  | US | USD | 18 |
|  1 | 2021 | UK    | GBP |       4 |

- `write_async` and `write_async_dataframe` functions with `max_workers` argument. 
This is essentially an easy way to speed up TM1 writeback through parallelization.

``` python
from TM1py import TM1Service

with TM1Service(address="", port=12354, ssl=True, user="admin", password="apple") as tm1:
    cells = {
        ('e1', 'e1'): 1,
        ('e1', 'e2'): 2,
        ('e2', 'e1'): 3,
        ('e2', 'e2'): 4,
        ('e3', 'e1'): 5,
        ('e3', 'e2'): 6,
    }

    tm1.cells.write_async(cube_name="c1", cells=cells, slice_size=1, max_workers=6)

```

``` python
from pandas import DataFrame

from TM1py import TM1Service

with TM1Service(address="", port=12354, ssl=True, user="admin", password="apple") as tm1:
    data = DataFrame({
        'd1': ['e1', 'e1', 'e2', 'e2', 'e3', 'e3'],
        'd2': ['e1', 'e2', 'e1', 'e2', 'e1', 'e2'],
        'value': [1, 2, 3, 4, 5, 6],
    })

    tm1.cells.write_dataframe_async(cube_name="c1", data=data, slice_size_of_dataframe=1, max_workers=6)

```

- TM1 git support #447 
TM1py now allows you to manage the TM1 git repository. You can manage the TM1 database with git and TM1py. 

- New authentication mode based on `cam_passport`. Can be used from Jupyter within Cognos Analytics.

``` python
from ca_data_connector.ca_lib import OnPremSession

session = OnPremSession()
session.initialize()
passport = session.get_cookies()['cam_passport']

with TM1Service(address=address, port=port, ssl=ssl, cam_passport=passport) as tm1:
    print(tm1.server.get_server_name())
```

# New Features

- AuditLog queries 3771dad
- Changes to `tm1.power_bi.execute_mdx` function to provide data to PBI in the most friendly way #476
- Improvement on PowerBI dimension retrieval function to fetch attribute for a parent instead of parent element name #478
- Support `MDX PROPERTIES` in `execute_mdx_dataframe` #470
- Introduce a new `cancel_at_timeout` argument to abort TM1 operation once the `timeout` is reached
- Support process debugging a83da11
- New arguments on `execute_mdx`, `execute_view`: `skip_cell_properties`, `element_unique_names` to control shape of response dictionary f751edf, 6d9e520
- TM1 git support #447 
- New `write_dataframe_async` function 94081eb
- New `write_async` function 3969c1d
- Create, read, update, delete servers on admin host #296 
- New `re_authenticate` function in TM1Service 2028d3d
- Better delta log requests aeed032
- Allow login with `cam_passport` f070c06

# Fixes

- Never attempt to delete leaves hierarchy a0aa152
- Make install with pandas work with zsh c458e40
- Fix encoding of ui-data property 6438309
- Fix deactivate / activate chore #515 
- Accept float as `timeout` 2d0c555
- Fix is_admin property in User class d9ed1b2
- Fix documentation on [readthedocs](https://tm1py.readthedocs.io/en/latest/)
- Remove linebreaks when writing with ti #568
- empty strings suppression in `extract_cellset_raw` 01eb4a7
- Allow parameters on unbound processes 144db38
- Improve `kwargs` delegation through call stack ee70f54
- Allow `WITH MEMBERS` combined with `include_attributes=True` in `execute_mdx_dataframe` function #593
- Allow `WITH MEMBERS` combined with `display_attribute=True` in `execute_mdx_dataframe_shaped` function #593
- Fixes #595 

# Acknowledgments

Big thanks to @lapstue, @kaleming, @beckyconning, @macsir, @jrobinsonAG, @tomasfelcman, @cwffonseca for contributing code to this release, and many others for reporting bugs and requesting features.

# How to upgrade TM1py

To upgrade TM1py, just use the following command:

```
pip install TM1py --upgrade
````

<hr /><em>This discussion was created from the release <a href='https://github.com/cubewise-code/tm1py/releases/tag/'>1.8.0</a>.</em>

**Comment by rkvinoth** [comment]
Glad to see the growth and new contributions. TM1py is definitely evolving!

**Reply by MariusWirtz** [reply]
Without your contributions and support, it would be evolving at a much slower pace! Thank you very much @rkvinoth

**Reply by rkvinoth** [reply]
Thanks to you too for the tremendous support with issues and in analyzing/merging the new PRs.

**Comment by wimgielis** [comment]
Good job all !




Op wo 29 sep. 2021 om 14:35 schreef Vinoth Kumar Ravi <
***@***.***>

> Thanks to you too for the tremendous support with issues and in
> analyzing/merging the new PRs.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cubewise-code/tm1py/discussions/619#discussioncomment-1402249>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEDHULPBAFVKGTKL3GC72BDUEMB2JANCNFSM5E7T5CTA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
-- 

------

Best regards / Beste groeten,

Wim Gielis
MS Excel MVP 2011-2014
https://www.wimgielis.com <http://www.wimgielis.be>

**Reply by TM1Fanboy** [reply]
Awesome. Maybe some day I will be a contributor myself. ;-)
